File: voxform-fastapi/routers/audio.py
"""Audio routes (authenticated) — mirrors internal/handlers/audio.go."""

import json
import os

from fastapi import APIRouter, Depends, File, Form, HTTPException, Request, UploadFile
from databases import Database

import config
from auth_utils import new_id, rand_hex
from deps import get_current_user, get_db
import job_queue as jobqueue

router = APIRouter(tags=["audio"])


def _ok(data):
    return {"success": True, "data": data}


def _err(status: int, msg: str):
    raise HTTPException(status_code=status, detail={"success": False, "message": msg})


@router.post("/audio/upload-slot")
async def create_slot(request: Request, claims: dict = Depends(get_current_user),
                      db: Database = Depends(get_db)):
    body = await request.json()
    response_id = body.get("responseId") or ""
    mime_type = body.get("mimeType") or ""

    upload_id = rand_hex(8)  # 16-char hex
    ext = ".wav" if "wav" in mime_type else ".webm"

    raw_dir = os.path.join(config.STORAGE_PATH, "audio", "raw")
    os.makedirs(raw_dir, exist_ok=True)
    upload_path = os.path.join(raw_dir, f"{upload_id}{ext}")
    upload_url = f"{config.STORAGE_URL}/audio/raw/{upload_id}{ext}"

    # Upsert audio_jobs
    existing = await db.fetch_one(
        "SELECT id FROM audio_jobs WHERE response_id = :rid", {"rid": response_id}
    )
    if existing:
        await db.execute(
            "UPDATE audio_jobs SET status = 'UPLOADING', wav_path = :p, updated_at = NOW() WHERE id = :id",
            {"p": upload_path, "id": existing["id"]},
        )
    else:
        await db.execute(
            "INSERT INTO audio_jobs (id, response_id, status, wav_path) VALUES (:id, :rid, 'UPLOADING', :p)",
            {"id": new_id(), "rid": response_id, "p": upload_path},
        )

    return _ok({"uploadId": upload_id, "uploadPath": upload_path, "uploadUrl": upload_url})


@router.post("/audio/upload/{upload_id}", status_code=201)
async def upload(
    upload_id: str,
    file: UploadFile = File(...),
    responseId: str = Form(...),
    clientQcResult: str = Form(None),
    claims: dict = Depends(get_current_user),
    db: Database = Depends(get_db),
):
    return await _do_upload(upload_id, file, responseId, clientQcResult, db)


@router.get("/audio/job/{response_id}")
async def job_status(response_id: str, claims: dict = Depends(get_current_user),
                     db: Database = Depends(get_db)):
    # Verify org access via JOIN
    row = await db.fetch_one(
        "SELECT r.id FROM responses r "
        "JOIN sessions se ON se.id = r.session_id "
        "JOIN surveys su ON su.id = se.survey_id "
        "WHERE r.id = :id AND su.org_id = :oid",
        {"id": response_id, "oid": claims["org_id"]},
    )
    if not row:
        _err(404, "response not found")

    job = await db.fetch_one("SELECT * FROM audio_jobs WHERE response_id = :id", {"id": response_id})
    if not job:
        return _ok({"status": "NOT_STARTED"})
    return _ok(dict(job))


# ── Shared upload logic (used by authenticated + public routers) ──────────────

async def _do_upload(upload_id: str, file: UploadFile, response_id: str,
                     client_qc_result: str | None, db: Database):
    # Determine file path from audio_jobs record
    job = await db.fetch_one("SELECT * FROM audio_jobs WHERE response_id = :rid", {"rid": response_id})
    if not job:
        raise HTTPException(404, detail={"success": False, "message": "audio job not found"})

    save_path: str = job["wav_path"] or ""
    if not save_path:
        ext = ".wav" if (file.content_type or "").find("wav") >= 0 else ".webm"
        raw_dir = os.path.join(config.STORAGE_PATH, "audio", "raw")
        os.makedirs(raw_dir, exist_ok=True)
        save_path = os.path.join(raw_dir, f"{upload_id}{ext}")

    # Write file (200 MB limit)
    content = await file.read(200 * 1024 * 1024)
    with open(save_path, "wb") as f:
        f.write(content)

    if client_qc_result:
        try:
            qc_json = json.loads(client_qc_result)
            await db.execute(
                "UPDATE responses SET qc_result = :qc, updated_at = NOW() WHERE id = :id",
                {"qc": json.dumps(qc_json), "id": response_id},
            )
        except Exception:
            pass

    await db.execute(
        "UPDATE audio_jobs SET status = 'QC_PENDING', wav_path = :p, updated_at = NOW() WHERE response_id = :rid",
        {"p": save_path, "rid": response_id},
    )

    await jobqueue.enqueue("audio:process", {"responseId": response_id})

    return {"success": True, "data": {
        "jobId": job["id"],
        "bytes": len(content),
        "filename": file.filename,
    }}

File: voxform-go/internal/handlers/audio.go
package handlers

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"

	"github.com/go-chi/chi/v5"

	"github.com/voxform/api/internal/db"
	"github.com/voxform/api/internal/middleware"
	"github.com/voxform/api/internal/models"
	"github.com/voxform/api/internal/queue"
)

type AudioHandler struct {
	db          *db.DB
	queue       *queue.Queue
	storagePath string
	storageURL  string
}

func NewAudioHandler(database *db.DB, q *queue.Queue, storagePath, storageURL string) *AudioHandler {
	return &AudioHandler{db: database, queue: q, storagePath: storagePath, storageURL: storageURL}
}

// CreateSlot registers an upload intent and returns upload details.
func (h *AudioHandler) CreateSlot(w http.ResponseWriter, r *http.Request) {
	var body struct {
		ResponseID string `json:"responseId"`
		Filename   string `json:"filename"`
		MimeType   string `json:"mimeType"`
	}
	if err := Decode(r, &body); err != nil || body.ResponseID == "" {
		BadRequest(w, "responseId is required"); return
	}

	ext := ".webm"
	if body.MimeType == "audio/wav" || body.MimeType == "audio/wave" {
		ext = ".wav"
	}

	uploadID := randHex(16)
	filename := uploadID + ext
	destPath := filepath.Join(h.storagePath, "audio", "raw", filename)
	if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
		InternalError(w, err); return
	}

	// Create or update AudioJob record
	var count int
	h.db.DB.Get(&count, `SELECT COUNT(*) FROM audio_jobs WHERE response_id = ?`, body.ResponseID)
	if count == 0 {
		h.db.Exec(`
			INSERT INTO audio_jobs (id, response_id, status, wav_path, provider, attempt_count, created_at, updated_at)
			VALUES (?, ?, 'UPLOADING', ?, 'GROQ', 0, NOW(), NOW())
		`, newID(), body.ResponseID, destPath)
	} else {
		h.db.Exec(`UPDATE audio_jobs SET status = 'UPLOADING', wav_path = ?, updated_at = NOW() WHERE response_id = ?`, destPath, body.ResponseID)
	}

	OK(w, map[string]any{
		"uploadId":   uploadID,
		"uploadPath": destPath,
		"uploadUrl":  fmt.Sprintf("%s/audio/raw/%s", h.storageURL, filename),
	})
}

// Upload receives the multipart audio file and enqueues processing.
func (h *AudioHandler) Upload(w http.ResponseWriter, r *http.Request) {
	uploadID := chi.URLParam(r, "uploadId")

	// 200MB max
	r.ParseMultipartForm(200 << 20)

	file, header, err := r.FormFile("file")
	if err != nil { BadRequest(w, "no file in request"); return }
	defer file.Close()

	responseID := r.FormValue("responseId")
	if responseID == "" { BadRequest(w, "responseId is required"); return }

	// Determine extension from mime type or filename
	ext := filepath.Ext(header.Filename)
	if ext == "" { ext = ".webm" }
	if header.Header.Get("Content-Type") == "audio/wav" { ext = ".wav" }

	destPath := filepath.Join(h.storagePath, "audio", "raw", uploadID+ext)
	if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
		InternalError(w, err); return
	}

	dst, err := os.Create(destPath)
	if err != nil { InternalError(w, err); return }
	defer dst.Close()

	written, err := io.Copy(dst, file)
	if err != nil { InternalError(w, err); return }

	// Update AudioJob with final path
	h.db.Exec(`UPDATE audio_jobs SET wav_path = ?, status = 'QC_PENDING', updated_at = NOW() WHERE response_id = ?`, destPath, responseID)

	// Parse optional client QC result
	if qcStr := r.FormValue("clientQcResult"); qcStr != "" {
		var qc map[string]any
		if json.Unmarshal([]byte(qcStr), &qc) == nil {
			qcJSON, _ := json.Marshal(qc)
			h.db.Exec(`UPDATE responses SET qc_result = ?, updated_at = NOW() WHERE id = ?`, string(qcJSON), responseID)
		}
	}

	// Enqueue processing
	h.queue.Enqueue("audio:process", map[string]any{"responseId": responseID})

	var job models.AudioJob
	h.db.DB.Get(&job, `SELECT * FROM audio_jobs WHERE response_id = ? LIMIT 1`, responseID)

	Created(w, map[string]any{
		"jobId":    job.ID,
		"bytes":    written,
		"filename": header.Filename,
	})
}

// JobStatus returns current processing status for polling.
func (h *AudioHandler) JobStatus(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r)
	responseID := chi.URLParam(r, "responseId")

	// Verify ownership
	var count int
	h.db.DB.Get(&count, `
		SELECT COUNT(*) FROM responses r
		JOIN sessions s ON s.id = r.session_id
		JOIN surveys sv ON sv.id = s.survey_id
		WHERE r.id = ? AND sv.org_id = ?
	`, responseID, claims.OrgID)
	if count == 0 { NotFound(w, "response not found"); return }

	job, err := models.GetAudioJob(h.db.DB, responseID)
	if err != nil {
		OK(w, map[string]any{"status": "NOT_STARTED", "responseId": responseID}); return
	}
	OK(w, job)
}
